### PR TITLE
  Implement fixes in Raygun4JS to correctly handle XHRs and virtual pages + enables tests

### PR DESCRIPTION
**Pull request by darcythomas**
This PR fixes:
 1. A bug where under some rare race conditions xhr requests get attributed to a different parent resource (page url)
 2. It also updates and allows the unit tests for this project, to be run under a GitHub Action

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,505 +1,689 @@
...
+
+-->
+
+## [Unreleased]
+
+## v2.26.1 - 2023-07-12
+
+### Fixed
+
+- Fixed timing bug, where  some xhr requests may rarely be mis-attributed to a different page from which they were started on.
```
Comment (Olwiba) [inline CHANGELOG.md:40]:
```suggestion
- Fixed timing bug, where some xhr requests may rarely be mis-attributed to a different page from which they were started on.
```

File: jasmine.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "spec_dir": "src",
+  "spec_dir":  "src",
```
Comment (Olwiba) [inline jasmine.json:2]:
```suggestion
  "spec_dir": "src",
```

File: package.json [hunk 3]
```diff
@@ -19,26 +19,25 @@
...
-    "@wdio/static-server-service": "^6.0.16",
-    "@wdio/sync": "^6.1.8",
-    "chromedriver": "^103.0.0",
-    "cross-env": "^7.0.2",
-    "grunt": "^1.1.0",
-    "grunt-contrib-clean": "^2.0.0",
-    "grunt-contrib-concat": "^1.0.1",
-    "grunt-contrib-jshint": "^2.1.0",
-    "grunt-contrib-uglify": "^4.0.1",
+    "@wdio/cli": "^8.11.2",
```
Comment (Olwiba) [inline package.json:22]: What led to all these dependencies needing to be updated?

File: package.json [hunk 4]
```diff
@@ -50,5 +49,9 @@
     "tracking",
     "raygun",
     "clientside"
-  ]
+  ],
+  "dependencies": {
+    "@wdio/codemod": "^0.12.0",
+    "jscodeshift": "^0.15.0"
```
Comment (Olwiba) [inline package.json:55]: all dependencies for the JS provider are going to be developer deps as these are used to build the project

File: src/raygun.js [hunk 5]
```diff
@@ -1023,25 +1023,9 @@ var raygunFactory = function(window, $, undefined) {
           _successCallback(xhr, url, data);
         }
       };
-    } else if (window.XDomainRequest) {
-      xhr.ontimeout = function() {
-        if (_enableOfflineSave) {
-          Raygun.Utilities.log('Raygun: saved locally');
-          offlineSave(url, data);
-        }
-      };
```
Comment (Olwiba) [inline src/raygun.js:1032]: please explain this regression?

File: src/raygun.js [hunk 6]
```diff
@@ -1023,25 +1023,9 @@ var raygunFactory = function(window, $, undefined) {
...
-        Raygun.Utilities.log('posted to Raygun');
-
-        sendSavedErrors();
-        callAfterSend(this);
-
-        if (_successCallback && typeof _successCallback === 'function') {
-          _successCallback(xhr, url, data);
-        }
-      };
-    }
```
Comment (Olwiba) [inline src/raygun.js:1044]: this also appears to be a regression?

File: src/raygun.rum/index.js [hunk 7]
```diff
@@ -563,7 +563,7 @@ var raygunRumFactory = function (window, $, Raygun) {
 
         self.offset = i;
 
-        if (this._captureMissingRequests) {
+        if (self._captureMissingRequests) {
```
Comment (Olwiba) [inline src/raygun.rum/index.js:566]: nice 👍

File: tests/fixtures/common/instrumentXHRs.js [hunk 8]
```diff
@@ -15,7 +15,7 @@
     }
     this.__headers[arguments[0]] = arguments[1];
 
-    origSetRequestHeader.apply(this, arguments);
+     origSetRequestHeader.apply(this, arguments);
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:18]:
```suggestion
    origSetRequestHeader.apply(this, arguments);
```

File: tests/fixtures/common/instrumentXHRs.js [hunk 9]
```diff
@@ -36,11 +36,11 @@
       url: arguments[1]
     });
 
-    this.addEventListener('load', function() {
+     this.addEventListener('load', function() {
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:39]:
```suggestion
    this.addEventListener('load', function() {
```

File: tests/fixtures/common/instrumentXHRs.js [hunk 10]
```diff
@@ -36,11 +36,11 @@
...
         window.__completedXHRs.push(this);
     });
 
-    origOpen.apply(this, arguments);
+     origOpen.apply(this, arguments);
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:43]:
```suggestion
    origOpen.apply(this, arguments);
```

File: tests/fixtures/common/instrumentXHRs.js [hunk 11]
```diff
@@ -52,10 +52,10 @@
 
     window.__sentXHRs.push({
       xhr: this,
-      clientIp: this.getRequestHeader('X-Remote-Address')
+      clientIp:  this.getRequestHeader('X-Remote-Address')
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:55]:
```suggestion
      clientIp: this.getRequestHeader('X-Remote-Address')
```

File: tests/fixtures/common/instrumentXHRs.js [hunk 12]
```diff
@@ -52,10 +52,10 @@
...
     });
 
-    origSend.apply(this, arguments);
+     origSend.apply(this, arguments);
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:58]:
```suggestion
    origSend.apply(this, arguments);
```

File: tests/fixtures/common/instrumentXHRs.js [hunk 13]
```diff
@@ -77,13 +77,13 @@
         window.__completedXHRs.push(this);
       };
 
-      origXOpen.apply(this, arguments);
+       origXOpen.apply(this, arguments);
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:80]:
```suggestion
      origXOpen.apply(this, arguments);
```

File: tests/fixtures/common/instrumentXHRs.js [hunk 14]
```diff
@@ -77,13 +77,13 @@
...
     };
 
     XDomainRequest.prototype.send = function() {
       window.__requestPayloads.push(JSON.parse(arguments[0]));
 
-      origXSend.apply(this, arguments);
+       origXSend.apply(this, arguments);
```
Comment (Olwiba) [inline tests/fixtures/common/instrumentXHRs.js:86]:
```suggestion
      origXSend.apply(this, arguments);
```

File: tests/fixtures/v1/manualSend.html [hunk 15]
```diff
@@ -2,12 +2,12 @@
 <html>
   <head>
     <title>Raygun4JS with V1 API</title>
-    <script src="/fixtures/common/instrumentXHRs.js"></script>
+    <script src="./../common/instrumentXHRs.js"></script>
     <script src="/dist/raygun.js"></script>
   </head>
   <body>
     <script type="text/javascript">
-      Raygun.init('abcdef==', {allowInsecureSubmissions: true});
```
Comment (Olwiba) [inline tests/fixtures/v1/manualSend.html:10]: How come we needed to disable `allowInsecureSubmissions` ?
will this lead to some headaches down the line if the test runs it over HTTP?

File: tests/fixtures/v2/rumFetchStatusCodes.html [hunk 16]
```diff
@@ -17,19 +17,18 @@
...
           window.supportsFetch = true;
 
-          fetch('rumXhrStatusCodes.html');
-          fetch('rumXhrStatusCodes.html?foo=bar');
-          fetch('foo.html');
-          fetch('http://localhost:4567/fixtures/v2/rumXhrStatusCodes.html');
+          await   fetch('rumXhrStatusCodes.html');
+          await   fetch('rumXhrStatusCodes.html?foo=bar');
+          await   fetch('foo.html');
+          await   fetch('http://localhost:4567/fixtures/v2/rumXhrStatusCodes.html');
```
Comment (Olwiba) [inline tests/fixtures/v2/rumFetchStatusCodes.html:31]:
```suggestion
          await fetch('rumXhrStatusCodes.html');
          await fetch('rumXhrStatusCodes.html?foo=bar');
          await fetch('foo.html');
          await fetch('http://localhost:4567/fixtures/v2/rumXhrStatusCodes.html');
```

File: tests/fixtures/v2/rumReferencedFetchWithFetchSnippet.html [hunk 17]
```diff
@@ -73,19 +73,18 @@
...
           window.supportsFetch = true;
 
-          referencedFetch('rumXhrStatusCodes.html');
-          referencedFetch('rumXhrStatusCodes.html?foo=bar');
-          referencedFetch('foo.html');
-          referencedFetch('http://localhost:4567/fixtures/v2/rumXhrStatusCodes.html');
+        await  referencedFetch('rumXhrStatusCodes.html');
+        await  referencedFetch('rumXhrStatusCodes.html?foo=bar');
+        await  referencedFetch('foo.html');
+        await  referencedFetch('http://localhost:4567/fixtures/v2/rumXhrStatusCodes.html');
```
Comment (Olwiba) [inline tests/fixtures/v2/rumReferencedFetchWithFetchSnippet.html:87]:
```suggestion
        await referencedFetch('rumXhrStatusCodes.html');
        await referencedFetch('rumXhrStatusCodes.html?foo=bar');
        await referencedFetch('foo.html');
        await referencedFetch('http://localhost:4567/fixtures/v2/rumXhrStatusCodes.html');
```

File: tests/fixtures/v1/manualSend.html (hunk 15 above)
Comment (Olwiba) [reply tests/fixtures/v1/manualSend.html:10]: I've noticed this approach on the other files - so if changes are needed here then we should also remember to update the other files

File: tests/specs/v1/eventsXhr.js [hunk 18]
```diff
@@ -8,21 +8,21 @@ describe("XHR functional tests for /events with V1", function() {
...
-    browser.pause(6000);
+    await browser.pause(6000);
 
-    var inFlightXhrs = browser.execute(function () {
+    var inFlightXhrs = await browser.execute(function () {
       return window.__inFlightXHRs;
     });
 
-    var didPerformRequest = _.any(inFlightXhrs, function (req) {
+    var didPerformRequest =  _.any(inFlightXhrs, function (req) {
```
Comment (Olwiba) [inline tests/specs/v1/eventsXhr.js:20]:
```suggestion
    var didPerformRequest = _.any(inFlightXhrs, function (req) {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 19]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
-    browser.pause(4000);
+    await browser.pause(4000);
 
-    var requestPayloads = browser.execute(function () {
+    var requestPayloads = await browser.execute(function () {
       return window.__requestPayloads;
     });
 
-    var passes = _.any(requestPayloads, function (payload) {
+    var passes =  _.any(requestPayloads, function (payload) {
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:108]:
```suggestion
    var passes = _.any(requestPayloads, function (payload) {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 20]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:87]:
```suggestion
    var passes = _.any(requestPayloads, function (payload) {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 21]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
       return payload.Details.Tags[0] === 'my_tag';
     });
 
-    expect(passes).toBe(true);
+     expect(passes).toBeTrue();
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:112]:
```suggestion
     expect(passes).toBeTrue();
```
```suggestion
    expect(passes).toBeTrue();
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 22]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
-    browser.pause(4000);
+    await browser.pause(4000);
 
-    var requestPayloads = browser.execute(function () {
+    var requestPayloads = await browser.execute(function () {
       return window.__requestPayloads;
     });
 
-    var passes = _.any(requestPayloads, function (payload) {
+    var passes =  _.any(requestPayloads, function (payload) {
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:126]:
```suggestion
    var passes = _.any(requestPayloads, function (payload) {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 23]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
       return payload.Details.UserCustomData.myCustomKey === 'myCustomValue';
     });
 
-    expect(passes).toBe(true);
+     expect(passes).toBeTrue();
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:130]:
```suggestion
    expect(passes).toBeTrue();
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 24]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
-    browser.pause(4000);
+    await browser.pause(4000);
 
-    var requestPayloads = browser.execute(function () {
+    var requestPayloads = await browser.execute(function () {
       return window.__requestPayloads;
     });
 
-    var passes = _.any(requestPayloads, function (payload) {
+    var passes =  _.any(requestPayloads, function (payload) {
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:144]:
```suggestion
    var passes = _.any(requestPayloads, function (payload) {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 25]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
       return payload.Details.User.Identifier === 'user_email_address@localhost.local' &&
         payload.Details.User.IsAnonymous === false &&
         payload.Details.User.FirstName === 'Foo' &&
         payload.Details.User.FullName === 'Foo Bar' &&
         payload.Details.User.UUID === 'BAE62917-ACE8-ab3D-9287-B6A33B8E8C55';
     });
 
-    expect(passes).toBe(true);
+     expect(passes).toBeTrue();
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:152]:
```suggestion
    expect(passes).toBeTrue();
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 26]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
   });
 
-  it("has correct version in payload when Raygun.setVersion() is called", function () {
+  it("has correct version in payload when Raygun.setVersion() is called",async function () {
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:155]:
```suggestion
  it("has correct version in payload when Raygun.setVersion() is called", async function () {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 27]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
-    browser.pause(4000);
+    await browser.pause(4000);
 
-    var requestPayloads = browser.execute(function () {
+    var requestPayloads = await browser.execute(function () {
       return window.__requestPayloads;
     });
 
-    var passes = _.any(requestPayloads, function (payload) {
+    var passes =  _.any(requestPayloads, function (payload) {
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:166]:
```suggestion
    var passes = _.any(requestPayloads, function (payload) {
```

File: tests/specs/v1/payloadManualSendTests.js [hunk 28]
```diff
@@ -7,167 +7,167 @@ var _entriesEndpoint = 'https://api.raygun.io/entries';
...
       return payload.Details.Version === '1.0.0.0';
     });
 
-    expect(passes).toBe(true);
+     expect(passes).toBeTrue();
```
Comment (Olwiba) [inline tests/specs/v1/payloadManualSendTests.js:170]:
```suggestion
    expect(passes).toBeTrue();
```

File: tests/specs/v2/clientIp.js [hunk 29]
```diff
@@ -4,35 +4,35 @@ var _ = require('underscore');
...
+    await expect(remoteAddressIsUndefined).toBe(true);
   });
 
-  it("X-Remote-Address is equal to '192.168.0.12'", function () {
-    browser.url('http://localhost:4567/fixtures/v2/withClientIpSet.html');
+  it("X-Remote-Address is equal to '192.168.0.12'", async function () {
+    await browser.url('http://localhost:4567/fixtures/v2/withClientIpSet.html');
 
-    browser.pause(4000);
+    await browser.pause(9000);
```
Comment (Olwiba) [inline tests/specs/v2/clientIp.js:26]: has the async behaviour caused longer test times?

File: tests/specs/v2/requestId.js [hunk 30]
```diff
@@ -18,6 +18,6 @@ describe("Request ID tests", function() {
     var requestId1 = requestPayloads[1].eventData[0].requestId;
     var requestId2 = requestPayloads[2].eventData[0].requestId;
 
-    expect(requestId1).not.toMatch(requestId2);
+    expect(requestId1 !== requestId2).toBeTrue(requestId2);
```
Comment (Olwiba) [inline tests/specs/v2/requestId.js:21]: why this change? can't we use nice jasmine test comparators here?

File: tests/specs/v2/rumXhrStatusTracking.js [hunk 31]
```diff
@@ -4,8 +4,16 @@ var _ = require('underscore');
 
 describe("RUM status code tracking", function() {
 
-  afterEach(function() {
-    browser.reloadSession();
+  beforeEach(function() {
+    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;
```
Comment (Olwiba) [inline tests/specs/v2/rumXhrStatusTracking.js:9]: what is the test execution time impact of this change?

File: tests/specs/v2/rumXhrStatusTracking.js [hunk 32]
```diff
@@ -42,75 +52,71 @@ describe("RUM status code tracking", function() {
       var pairStatus = expectedPairs[i].statusCode;
       var pairType = expectedPairs[i].type;
 
-      expect(payloadUrl).toBe(pairUrl, "failed for type: " + pairType);
-      expect(payloadStatus).toBe(pairStatus, "failed for type: " + pairType);
-      expect(payloadDataType).toBe(payloadDataType, "XHR data type missing for: " + pairType);
+      expect(payloadUrl === pairUrl).toBeTrue("failed for type: " + pairType);
+      expect(payloadStatus === pairStatus).toBeTrue( "failed for type: " + pairType);
+      expect(payloadDataType === payloadDataType).toBeTrue( "XHR data type missing for: " + pairType);
```
Comment (Olwiba) [inline tests/specs/v2/rumXhrStatusTracking.js:57]: we have lost the jasmine comparitors here?

File: tests/specs/v2/rumXhrStatusTracking.js [hunk 33]
```diff
@@ -42,75 +52,71 @@ describe("RUM status code tracking", function() {
...
+    await checkStatusCodes();
   });
 
-  it("attaches status codes to requests for fetch requests", function() {
-    browser.url('http://localhost:4567/fixtures/v2/rumFetchStatusCodes.html');
+  it("attaches status codes to requests for fetch requests", async function() {
+    await browser.url('http://localhost:4567/fixtures/v2/rumFetchStatusCodes.html');
 
-    browser.pause(1000);
+    await browser.pause(40000);
```
Comment (Olwiba) [inline tests/specs/v2/rumXhrStatusTracking.js:73]: 1s to 40s here?

File: tests/specs/v2/rumXhrStatusTracking.js [hunk 34]
```diff
@@ -42,75 +52,71 @@ describe("RUM status code tracking", function() {
...
+    var supportsFetch = await browser.execute(function() {
       return window.supportsFetch;
     });
 
     if (!supportsFetch) {
       return;
     }
 
-    browser.pause(35000);
+   // await browser.pause(3000);
```
Comment (Olwiba) [inline tests/specs/v2/rumXhrStatusTracking.js:83]: not needed?

File: tests/specs/v2/rumXhrStatusTracking.js [hunk 35]
```diff
@@ -127,17 +133,17 @@ describe("RUM status code tracking", function() {
 
         for (var i = 0; i < expectedCalls.length; i++) {
           var url = expectedCalls[i];
-          expect(completedCalls.indexOf(url)).not.toBe(-1);
+          expect(completedCalls.indexOf(url) !== -1).toBeTrue();
```
Comment (Olwiba) [inline tests/specs/v2/rumXhrStatusTracking.js:136]: jasmine competitors are gone

**Review by Olwiba — CHANGES_REQUESTED** [review]
Have completed my review of the changes. I'm a little concerned about the test timing blowing out or becoming flakey due to the async behaviour implemented, how come we added all this? Is it just to get GH actions to work? 

Otherwise just some minor code formatting changes, I wonder if there is a `eslint` file in this project to standardise how IDEs apply code styles

File: package.json (hunk 3 above)
Comment (darcythomas) [reply package.json:22]: Well!

I needed to update the node version to one which was supported:
1. Period
2. On My machine
3. On Github actions 

I then had to update the testing libraries etc. So that would compile under a supported version of node.

Because of a breaking changes in the wdio & jasmine i needed to update thew tests (`async` and removing .`not.` respectively)

File: src/raygun.js (hunk 5 above)
Comment (darcythomas) [reply src/raygun.js:1032]: Fair.

This was a missed piece for another PR.
`window.XDomainRequest` has been obsoleted since IE8.

So I removed it from here as well.

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 32 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js:57]: Yes. It throws a 'Negation comparison not an option' (or some thing like that).

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 31 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js:9]: Not much this only applies for these few tests. and most of them complete will within the timeout period.

It takes about 5 min to complete all of the tests

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 34 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js:83]: No it is not doing anything.
```suggestion
   
```

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 35 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js:136]: Yes <sad face>

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 33 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js:73]: Yes. for some tests it needed to be 40s, to complete reliably

File: package.json (hunk 4 above)
Comment (darcythomas) [reply package.json]: Cleaned up

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 31 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js:9]: To took about 5 min the last time these tests were run successfully

Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js]: Ok i have made this only pause as long as it needs to (up to 40 seconds)

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 33 above)
Comment (darcythomas) [reply tests/specs/v2/rumXhrStatusTracking.js]: Made this better that it will only pause as long as it needs to (up to 40 seconds)

File: tests/specs/v2/rumXhrStatusTracking.js (hunk 31 above)
Comment (Olwiba) [reply tests/specs/v2/rumXhrStatusTracking.js:10]: sounds good, seeing as these are only being run by GH actions this won't have an impact on any TC builds. Happy for the added test time due to the async behaviour 😄

**Review by Olwiba — APPROVED** [review]
Outstanding work on this one Darcy! Got some good wins in this PR for the provider 😄 
Hopefully, you enjoyed working in the JS space..

Let me know how the pre-soak goes and if there's anything I can do to help with testing.
